src/js/actions/layers.js

```diff
@@ -134,6 +132,135 @@ define(function (require, exports) {
+            }
+        });
+        
+        var essentialPropertiesPromise = essentialDescriptorCommands.length === 0 ? Promise.resolve([]) :
+                descriptor.batchPlay(essentialDescriptorCommands),
+            selectedPropertiesPromise = selectedDescriptorCommands.length === 0 ? Promise.resolve([]) :
+                descriptor.batchPlay(selectedDescriptorCommands);
+
+        var extensionPromise = Promise.resolve([]);
+        if (options.getExtensionData) {
+            var extensionPlayObjects = layers.map(function (layer) {
+                var layerRef = layerLib.referenceBy.id(layer.id);
+
+                return layerLib.getExtensionData(docRef, layerRef, METADATA_NAMESPACE);
+            });
+
+            extensionPromise = descriptor.batchPlayObjects(extensionPlayObjects)
+                .map(function (extensionData) {
+                    var extensionDataRoot = extensionData[METADATA_NAMESPACE];
+                    return (extensionDataRoot && extensionDataRoot.exportsMetadata) || {};
+                });
+        }
+        
+
+        return Promise.join(
+            essentialPropertiesPromise,
+            selectedPropertiesPromise,
+            extensionPromise,
+            function (essential, selected, extension) {
+                selected.forEach(function (properties) {
+                    properties.initialized = true;
+                });
+                
+                if (extension.length !== 0) {
+                    layers.forEach(function (layer, index) {
+                        extension[index].layerID = layer.id;
+                    });
+                }
+                
+                var boundsDescriptorCommands = [];
+                
+                if (options.getBoundsProperties) {
+                    var essentialDescriptors = essential;
+                    
+                    if (!options.getEssentialProperties) {
+                        essentialDescriptors = layers.map(function (layer) {
+                            return layer.descriptor.toJS();
+                        });
+                    }
+                    
+                    essentialDescriptors.forEach(function (d) {
+                        var requiredBounds = Bounds.getRequiredBoundsName(d),
+                            layerRef = layerLib.referenceBy.id(d.layerID);
+                            
+                        if (requiredBounds) {
+                            boundsDescriptorCommands.push(multiGetRef(docRef, layerRef, ["layerID", requiredBounds]));
+                        }
+                    });
+                }
+
+                var boundsPromise = boundsDescriptorCommands.length === 0 ? Promise.resolve([]) :
+                        descriptor.batchPlay(boundsDescriptorCommands);
+
+                return boundsPromise.then(function (bounds) {
+                    bounds.forEach(function (properties) {
+                        properties.boundsInitialized = true;
```

Comment:
Mark the `descriptor.boundsInitialized` so that we can know from the Layer model whether its bounds is ready or not.